### DISCUSSION TITLE
Value of indicator for different Instrument

### DISCUSSION THREAD

**Discussion by rsmb7z** [nautechsystems/nautilus_trader#844] [Q&A] [answered]
Hi @cjdsellers 
I have a scenario to know the indicator's value of another instrument for example while in stock strategy knowing the SMA of SPY.
Is there built-in way to achieve this? If not can you please give your thoughts what could be good approach to achieve this. One of the option I think may work is using signals as specified in SignalStrategy from examples.

**Comment by cjdsellers** [comment]
Hi @rsmb7zhx 

I'd recommend reading up a little more on strategies, actors and indicators. 

For this one I'd probably just instantiate the built in SMA indicator inside your strategy (see the example strategies using the EMA), and registering it with the data feed you're interested in taking the average of, so I'm not sure if you're after an average of trades, or bars etc

Hope that helps!

**Reply by rsmb7z** [reply]
Hi @cjdsellers 
I have tested the built-in `subscribe_bars` which works perfectly, but only for the same Instrument as of strategy. If you subscribe to BarType other than the Instrument of the strategy, it would simply get no data. Because the strategy's `on_bar` does not receives Bars, of other instrument, indicators for it doesn't get initialized as well.

**Comment by cjdsellers** [comment, marked as answer]
Any strategy can trade any number of instruments, and subscribe to any data though. 

Could you please share some code snippets so we can see whats going on?

**Reply by rsmb7z** [reply]
Hi @cjdsellers 
To reproduce steps as follows (any instruments can be used):
1. Setup Engine so it has Bars data loaded for more than one instrument i.e. 'EUR/USD.IDEALPRO' and 'GBP/USD.IDEALPRO'
2. Modify `EMACrossBracket` from examples and add following `on_start`
```python
        self.subscribe_bars(BarType.from_str("EUR/USD.IDEALPRO-1-MINUTE-BID-EXTERNAL"))
        self.subscribe_bars(BarType.from_str("GBP/USD.IDEALPRO-1-MINUTE-BID-EXTERNAL"))
```
4. Set the instrument_id for EMACrossBracketConfig to 'EUR/USD.IDEALPRO'
6. Run the strategy, console will print received bars [confirming 'on_bar' for only 'EUR/USD' is triggered] and not any other instrument
7. Now change the instrument_id (in step 4) to EMACrossBracketConfig to 'GBP/USD.IDEALPRO'
8. Run the strategy again, console will print received bars [confirming 'on_bar' for only 'GBP/USD' is triggered] and not any other instrument

So you can subscribe BarType of different instrument, there is no error but it will not receive the data either. Not sure if the behavior is intentional or a bug.

Also as you can see in the EMACrossBracketConfig from examples, strategy is bound to single instrument string and doesn't accept a list. I believe this behavior makes sense, one-strategy should trade one-instrument, but allowing to subscribe data for another instrument is a core of strategy's decision making logic.
As per documentation/thread [I cannot recall where], Strategy can be reused for different instrument by adding unique `order_id_tag`, that's what I understood.

```python
# Configure your strategy
    config = EMACrossBracketConfig(
        instrument_id=str(GBPUSD_SIM.id),
        bar_type="GBP/USD.SIM-1-MINUTE-BID-EXTERNAL",
        fast_ema_period=10,
        slow_ema_period=20,
        bracket_distance_atr=3.0,
        trade_size=Decimal(1_000_000),
    )
```

**Reply by cjdsellers** [reply]
I just setup a quick backtest and confirmed you can receive data from multiple instruments (I defined separated instruments but passed the same CSV data to the wranglers.

```
2020-08-14T14:24:35.816999936Z [INF] BACKTESTER-001.EMACross-000: Bar(BTCUSDT.BINANCE-250-TICK-LAST-INTERNAL,424.20,424.73,424.04,424.38,452.560000,1597415075816999936)
2020-08-14T14:26:22.517999872Z [INF] BACKTESTER-001.EMACross-000: Bar(ETHUSDT.BINANCE-250-TICK-LAST-INTERNAL,424.38,425.21,424.23,425.21,635.74969,1597415182517999872)
2020-08-14T14:26:22.517999872Z [INF] BACKTESTER-001.EMACross-000: Bar(BTCUSDT.BINANCE-250-TICK-LAST-INTERNAL,424.38,425.21,424.23,425.21,635.749690,1597415182517999872)
2020-08-14T14:27:40.157999872Z [INF] BACKTESTER-001.EMACross-000: Bar(ETHUSDT.BINANCE-250-TICK-LAST-INTERNAL,425.21,425.22,424.28,424.30,717.92037,1597415260157999872)
```

When a strategy subscribes to data, it will subscribe to the message bus, and then send a subscribe command to the data engine:
```python
        self._msgbus.subscribe(
            topic=f"data.bars.{bar_type}",
            handler=self.handle_bar,
        )

        cdef Subscribe command = Subscribe(
            client_id=client_id,
            venue=bar_type.instrument_id.venue,
            data_type=DataType(Bar, metadata={"bar_type": bar_type}),
            command_id=UUID4(),
            ts_init=self._clock.timestamp_ns(),
        )
```

So any message received on the `topic` will be passed to the `handler`, its core functionality for the platform.

For this case of the `EMACross` strategy, there must be something going wrong either:
- Not all data has been added
- The logic of the strategy is not subscribing to the intended data (although the above does look correct)

The example strategy does take a single instrument, but this is just an example after all. You could try modifying the config to take several or a list of instruments also.

**Reply by rsmb7z** [reply]
Hi @cjdsellers 
Thanks for your support and time. It works awsome.
The issue was with my config as loading through BacktestRunConfig. So either `instrument_id` parameter needs to be removed from data config section or add the config for each of the required instruments. Thanks again.
